src/GitHub.Api/Git/Repository.cs

```diff
@@ -149,30 +160,151 @@ private void RepositoryManager_OnStatusUpdated(GitStatus status)
             CurrentStatus = status;
         }
 
-        private void RepositoryManager_OnActiveRemoteChanged(ConfigRemote? remote)
+        private void RepositoryManager_OnLocksUpdated(IEnumerable<GitLock> locks)
         {
-            CurrentRemote = remote;
+            CurrentLocks = locks;
+            OnLocksChanged?.Invoke(CurrentLocks);
         }
 
-        private void RepositoryManager_OnActiveBranchChanged(ConfigBranch? branch)
+        private void RepositoryManager_OnHeadUpdated(string h)
         {
-            CurrentBranch = branch;
+            if (head != h)
+            {
+                head = h;
+                var branch = GetCurrentBranch();
+                var remote = GetCurrentRemote();
+
+                if (!Nullable.Equals(currentBranch, branch))
+                {
+                    currentBranch = branch;
+                    OnCurrentBranchChanged?.Invoke(currentBranch.HasValue ? currentBranch.Value.Name : null);
+                }
+
+                if (!Nullable.Equals(currentRemote, remote))
+                {
+                    currentRemote = remote;
+                    OnCurrentRemoteChanged?.Invoke(currentRemote.HasValue ? currentRemote.Value.Name : null);
+                }
+            }
         }
 
-        private void RepositoryManager_OnHeadChanged()
+        private void RepositoryManager_OnRemoteBranchListUpdated(Dictionary<string, Dictionary<string, ConfigBranch>> branches)
         {
-            OnHeadChanged?.Invoke();
+            remoteBranches = branches;
+            OnRemoteBranchListChanged?.Invoke();
```

Comment:
Currently we assume every update is a full change.
Should we compare and fire changed when appropriate?